sqlgen/db.go

```diff
@@ -100,35 +110,82 @@ func NewDB(conn *sql.DB, schema *Schema) *DB {
 // pairs.
 func (db *DB) WithShardLimit(shardLimit Filter) (*DB, error) {
 	if db.shardLimit != nil {
-		return nil, errors.New("already limited")
+		return nil, errors.New("already has shard limit")
 	}
 
 	dbCopy := *db
 	dbCopy.shardLimit = shardLimit
 	return &dbCopy, nil
 }
 
-func (db *DB) checkFilterAgainstShardLimit(filter Filter) error {
-	if db.shardLimit == nil {
-		return nil
+// WithDynamicLimit is similar to WithShardLimit except that it supports
+// dynamically computing what filter to enforce as a limit, based on a user-
+// provide callback. It may be used together with a shard limit.
+//
+// WithDynamicLimit takes two arguments, `getLimitFilter`, and `onError`,
+// both user-provided callbacks.
+//
+// Before making an SQL query, a Thunder DB handle with a dynamic limit
+// will call `getLimitFilter` with the request context and the name of
+// the table being queried. It will use the sqlgen filter returned by
+// the function as the limit to check for against the current query.
+// The underlying limit check is the same as that used in `WithShardLimit`.
+//
+// If the limit check fails, Thunder will call `onError` with the resulting
+// error object.  If `onError` itself returns an error, the query will fail.
```

Comment:
it took me a couple reads to understand what `onError` is for. consider a name like `shouldContinue` or similar?

Reply:
Going with `shouldContinueOnError`.  As a refresher, this can't just be a global bool setting because we'd like to be able to notify user code that their dynamic limit check failed.